### PR TITLE
Prep for release to maven

### PR DESCRIPTION
**Pull request by lbustelo**
This fixes https://issues.apache.org/jira/browse/TOREE-245.

This PR contains modifications to build to release JAR artifacts of Toree.

- [x] Fix POM of root project
- [x] Signing of release packages for download (http://blog.sonatype.com/2010/01/how-to-generate-pgp-signatures-with-maven/#.Vm9Km8q22-q)
- [x] Signing of JARs published to maven (http://blog.sonatype.com/2010/01/how-to-generate-pgp-signatures-with-maven/#.Vm9Km8q22-q)
- [x] Run rat
- [x] Build src, bin, jdoc for submodules to publish on maven
- [x] Build assembly used in pip distro
- [x] Publish to local maven to test
- [x] Publish to apache staging to test
- [x] Package Notice/Licence files in all JARS
- [x] Package Notice/Licence file in bin distro
- [x] Package Notice/Licence file in pip distro
- [x] Build comprehensive Notice/License file for bin/pip/assembly packages
- [x] Automate using Travis? (not possible to fully automate... will need to look at Jenkins)
- [x] Add disclaimer to release packages (not including JARS) (http://incubator.apache.org/guides/branding.html#disclaimers)
- [x] Rename bin package to contain the name 'incubating'
- [x] Any additional disclaimer due to LGPL dependency
- [x] Create a source distribution. Based on http://www.apache.org/dev/release-publishing.html#valid
- [x] Validate that we can publish to real maven repo
- [x] Rename referent to toree kernel in notebook to Apache Toree
- [x] Release notes

### CODE REVIEW

**Comment by lbustelo** [comment]
I'm adjusting the checklist above based on http://incubator.apache.org/guides/releasemanagement.html#check-list

**Comment by Lull3rSkat3r** [comment]
Uploading files for distribution: http://www.apache.org/dev/release.html#upload-scp
Signing releases: http://www.apache.org/dev/release-signing.html#basic-facts

**Comment by lbustelo** [comment]
I'm going to work on `Rename bin package to contain the name 'incubating'`

**Comment by lbustelo** [comment]
Adding item to list to `Rename referent to toree kernel in notebook to Apache Toree` based on http://incubator.apache.org/guides/branding.htm

**Comment by lbustelo** [comment]
I'm going to work on `Package Notice/Licence file in pip distro`

**Comment by lbustelo** [comment]
Based on http://incubator.apache.org/guides/releasemanagement.html#best-practice-license I'm creating an item to list the License of other bundled dependencies in our package. i.e. LGPL.

File: etc/pip_install/setup.py [hunk 1]
```diff
@@ -27,12 +27,22 @@
 
 setup_args = dict(
     name='toree',
-    author='Toree Development Team',
+    author='Apache Toree Development Team',
     author_email='user@toree.incubator.apache.org',
```
Comment (lresende) [inline etc/pip_install/setup.py]: Toree does not have an user list, right ?

File: etc/tools/.rat-excludes [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+NAMESPACE
+package-sparkR.sh
+\.lintr
+.*ivy.xml
+\.example-image
+\.ensime
+\.example-image
+\.jvmopts
+\.pairs
+\.travis.yml
```
Comment (lresende) [inline etc/tools/.rat-excludes]: This list is to broad, and it should only contain files that does not support adding the apache license header As an example, see how travis.yml does support it : https://github.com/apache/incubator-zeppelin/blob/master/.travis.yml

File: project/Common.scala [hunk 3]
```diff
@@ -152,9 +170,9 @@ object Common {
     unmanagedResourceDirectories in Test +=
       (baseDirectory in Build.root).value / "resources/test",
 
-    publishTo := Some("Spark Kernel Nexus Repo" at repoUrl),
-
-    credentials += Credentials("Sonatype Nexus Repository Manager", repoHost, repoUsername, repoPassword),
```
Comment (lresende) [inline project/Common.scala]: I would call this a different name, I actually got confused, thinking you were directly publishing to Sonatype, while you are just publishing to the Apache staging repository

File: Makefile [hunk 4]
```diff
@@ -15,9 +15,9 @@
 # limitations under the License
 #
 
-.PHONY: help clean clean-dist build dev test test-travis release pip-release bin-release dev-binder .binder-image
+.PHONY: help clean clean-dist build dev test test-travis release pip-release bin-release dev-binder .binder-image audit
 
-VERSION?=0.1.0.dev5
+VERSION?=0.1.0.dev5-incubating
```
Comment (lresende) [inline Makefile]: Does pip have the same notion of snapshot, where new 'deploy' overrides previous one ? If so, I would just call it 0.1.0.dev.incubating.

File: project/Common.scala (hunk 3 above)
Comment (lbustelo) [inline project/Common.scala]: Have not gotten to publish yet. Will change that soon.

File: Makefile (hunk 4 above)
Comment (lbustelo) [inline Makefile]: Pip wants a new version each time. We could separate the pip dev suffix from the snapshot suffix.

File: etc/pip_install/setup.py (hunk 1 above)
Comment (lbustelo) [inline etc/pip_install/setup.py]: Good catch... Will change to dev.

File: etc/tools/.rat-excludes (hunk 2 above)
Comment (Lull3rSkat3r) [inline etc/tools/.rat-excludes]: This list was based on Spark's [rat-excludes](https://github.com/apache/spark/blob/master/dev/.rat-excludes). I did not test the data files, so those are eligible to be removed. However, most of the others appear if they are not included. If they are eligible for a header we can add the headers and remove them from the list. However, there are some files which get generated (like *.iml and db.lck) and cannot be removed from this list.

**Comment by lbustelo** [comment]
@lresende What is the best way to create the aggregate LICENSE and NOTICE file? We are using sbt-assembly, but it does not seem to be doing all the dependencies. For example, our LGPL dependency does not ship with a LICENSE in the jar, so it does not get merge.

Questions are then?
How do you get a list of what packages we need to aggregate into the LICENSE and NOTICE files.

**Comment by lresende** [comment]
There is no magic, I guess it's a manual process, by looking into all dependencies being packaged on the distribution and create the LICENSE and NOTICE file. Some projects does it a little different, where they add the LICENSE files into a folder and either point to them or aggregate them.

**Comment by lbustelo** [comment]
It seems that spark has all of it already compiled in their top level LICENSE and NOTICE files https://github.com/apache/spark/blob/branch-1.0/LICENSE

**Comment by Lull3rSkat3r** [comment]
I have compiled this list of license information for our direct and transitive dependencies. @lbustelo, you should be able to find information in most of those links for the NOTICES file.

BSD-like
========
Direct dependencies
-------------------
org.clapper:classutil_2.10:1.0.3 -> http://software.clapper.org/javautil/#copyright-and-license
org.scala-lang:scala-reflect:2.10.4 -> http://www.scala-lang.org/license.html
org.scalamacros:quasiquotes_2.10:2.0.1 -> http://www.scala-lang.org/license.html

Transitive Dependencies
-----------------------
jline:jline:2.6 -> https://github.com/jline/jline2/blob/master/LICENSE.txt
org.clapper:grizzled-scala_2.10:1.1.2 -> http://software.clapper.org/grizzled-scala/#copyright-and-license
org.clapper:grizzled-slf4j_2.10:1.0.1 -> http://software.clapper.org/grizzled-slf4j/#copyright-and-license
org.scala-stm:scala-stm_2.10:0.7 -> https://nbronson.github.io/scala-stm/license.html
org.scalaz:scalaz-concurrent_2.10:7.1.2  -> https://github.com/scalaz/scalaz/blob/series/7.1.x/etc/LICENCE
org.scalaz:scalaz-core_2.10:7.1.2 -> https://github.com/scalaz/scalaz/blob/series/7.1.x/etc/LICENCE
org.scalaz:scalaz-effect_2.10:7.1.2 -> https://github.com/scalaz/scalaz/blob/series/7.1.x/etc/LICENCE
asm:asm-commons:3.3 -> http://asm.ow2.org/license.html
asm:asm-tree:3.3 -> http://asm.ow2.org/license.html
asm:asm:3.3 -> http://asm.ow2.org/license.html

MIT license
===========
Direct dependencies
-------------------
net.sf.jopt-simple:jopt-simple:4.6 -> https://pholser.github.io/jopt-simple/license.html
org.slf4j:slf4j-api:1.7.5 -> http://www.slf4j.org/license.html
org.slf4j:slf4j-api:1.7.1 -> http://www.slf4j.org/license.html

Transitive Dependencies
-----------------------
com.lihaoyi:ammonite-terminal_2.10:0.5.0 -> https://github.com/lihaoyi/Ammonite/blob/master/LICENSE

GNU General Lesser Public License (LGPL) version 3.0
====================================================
Direct dependencies
-------------------
org.zeromq:jeromq:0.3.4 -> https://github.com/zeromq/jeromq/blob/master/COPYING

**Comment by lbustelo** [comment]
a couple of links guiding NOTICE file content

http://www.apache.org/dev/licensing-howto.html#mod-notice

>Copyright notifications which have been relocated from source files (rather than removed) must be preserved in NOTICE. However, elements such as the copyright notifications embedded within BSD and MIT licenses need not be duplicated in NOTICE -- it suffices to leave those notices in their original locations.

>It is important to keep NOTICE as brief and simple as possible, as each addition places a burden on downstream consumers.

>Do not add anything to NOTICE which is not legally required.

http://apache.org/legal/resolved.html#required-third-party-notices

>When a release contains third party works, the licenses covering those works may ask that consumers are informed in certain specific fashions. These third party notices vary from license to license. Apache releases should contain a copy of each license, usually contained in the LICENSE document. For many licenses this is a sufficient notice. For some licenses some additional notice is required. In many cases, this will be included within the dependent artifact.

>A required third-party notice is any third party notice which isn't covered by the above cases.

>See Bundling Other ASF Products for a note on required notices when a release contains another Apache product.

**Comment by lbustelo** [comment]
based on https://github.com/zeromq/jeromq/blob/master/COPYING.LESSER#L91... I'm adding the license to the uber packages (pip, bin, and assembly jar)

**Comment by lbustelo** [comment]
Thread regarding single release with LGPL http://apache-incubator-general.996316.n3.nabble.com/Release-dependant-on-LGPL-td48424.html

**Comment by Lull3rSkat3r** [comment]
In case we need it, here is a list of all the Apache Dependencies:

commons-logging:commons-logging:1.1.3
org.apache.ivy:ivy:2.4.0-rc1
com.google.guava:guava:14.0.1
org.apache.commons:commons-exec:1.3
org.joda:joda-convert:1.6
joda-time:joda-time:2.3
com.github.alexarchambault:coursier-cache_2.10:1.0.0-M9
com.github.alexarchambault:coursier_2.10:1.0.0-M9
com.typesafe.akka:akka-slf4j_2.10:2.3.11
com.typesafe.akka:akka-actor_2.10:2.3.11
com.typesafe:config:1.2.1
com.typesafe.play:play-json_2.10:2.3.6
com.typesafe.play:play-datacommons_2.10:2.3.6
com.typesafe.play:play-functional_2.10:2.3.6
com.typesafe.play:play-iteratees_2.10:2.3.6
org.springframework:spring-core:4.1.1.RELEASE

**Comment by Lull3rSkat3r** [comment]
I have a [PR](https://github.com/lbustelo/incubator-toree/pull/11) for removing the systems tests. I created a [jira](https://issues.apache.org/jira/browse/TOREE-296) for adding them back with [jupyter kernel test](https://github.com/jupyter/jupyter_kernel_test).

**Comment by lbustelo** [comment]
@chipsenkbeil Please take a look at this. I think it should be ready for merge. We still do not have a way to automate some of this, but I think we can tackle that separately. There are quite a bit of changes here that it's better to merge now and than having to keep rebasing this PR.

**Comment by chipsenkbeil** [comment]
@lbustelo @Lull3rSkat3r, read through it all and looks good. It looks like the system tests have been removed (along with using a consistent Akka timeout). Do we have plans in the near future to use the Jupyter kernel test project?

File: kernel-api/src/test/scala/org/apache/toree/utils/TaskManagerSpec.scala [hunk 5]
```diff
@@ -34,7 +34,7 @@ class TaskManagerSpec extends FunSpec with Matchers with MockitoSugar
   with Eventually with Timeouts
 {
   implicit override val patienceConfig = PatienceConfig(
-    timeout = scaled(Span(200, Milliseconds)),
+    timeout = scaled(Span(30, Seconds)),
```
Comment (chipsenkbeil) [inline kernel-api/src/test/scala/org/apache/toree/utils/TaskManagerSpec.scala]: Should this have a constant under `test.utils` similar to the Akka one?

**Comment by lbustelo** [comment]
@chipsenkbeil There is a separate [issue](https://issues.apache.org/jira/browse/TOREE-296) for that.

File: kernel-api/src/test/scala/org/apache/toree/utils/TaskManagerSpec.scala (hunk 5 above)
Comment (lbustelo) [inline kernel-api/src/test/scala/org/apache/toree/utils/TaskManagerSpec.scala]: Maybe... my conversation with @Lull3rSkat3r was that I feel these tests are trying to be units, but are actually more of an integration test. Either we find a different way to test the functionality without involving Akka, or we moved them as part of overall SystemTest.

**Comment by chipsenkbeil** [comment]
Cool, then LGTM. Want me to merge?

**Comment by lbustelo** [comment]
@chipsenkbeil yes... merge it

**Pull request merged**